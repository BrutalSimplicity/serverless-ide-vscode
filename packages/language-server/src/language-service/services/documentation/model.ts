export const LIST_TYPE: 'List' = 'List';
export const MAP_TYPE: 'Map' = 'Map';
export const STRING_TYPE = 'String';
export const INTEGER_TYPE = 'Integer';
export const BOOLEAN_TYPE = 'Boolean';

export type PrimitiveType =
	| typeof STRING_TYPE
	| typeof INTEGER_TYPE
	| typeof BOOLEAN_TYPE;

export type ItemType = {
	Type?: typeof LIST_TYPE | typeof MAP_TYPE | string;
	ItemType?: string;
	PrimitiveType?: PrimitiveType;
	PrimitiveItemType?: typeof STRING_TYPE | typeof INTEGER_TYPE;
};

export type Property = {
	Documentation: string;
	Required: boolean;
	UpdateType?: 'Mutable' | 'Immutable' | 'Conditional';
	DuplicatesAllowed?: boolean;
} & ItemType;

export type Attribute = ItemType;

export interface PropertySpecification {
	Documentation: string;
	Properties: {
		[key: string]: Property;
	};
}

export interface ResourceSpecification {
	Documentation: string;
	Attributes: {
		[key: string]: Attribute;
	};
	Properties: {
		[key: string]: Property;
	};
}

export interface Specification {
	PropertyTypes: {
		[key: string]: PropertySpecification;
	};
	ResourceTypes: {
		[key: string]: ResourceSpecification;
	};
}
