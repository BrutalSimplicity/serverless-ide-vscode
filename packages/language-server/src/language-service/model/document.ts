export enum DocumentType {
	CLOUD_FORMATION = "CLOUD_FORMATION",
	SAM = "SAM",
	SERVERLESS_FRAMEWORK = "SERVERLESS_FRAMEWORK",
	UNKNOWN = "UNKNOWN"
}
