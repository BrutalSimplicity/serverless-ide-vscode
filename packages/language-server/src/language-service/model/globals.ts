import { Segment } from "vscode-json-languageservice"

export const API: "Api" = "Api"
export const FUNCTION: "Function" = "Function"
export const SIMPLE_TABLE: "SimpleTable" = "SimpleTable"

export type GlobalKeyType = typeof API | typeof FUNCTION | typeof SIMPLE_TABLE

export const GlobalKeys = {
	[API]: {
		resourceType: "AWS::Serverless::Api"
	},
	[FUNCTION]: {
		resourceType: "AWS::Serverless::Function"
	},
	[SIMPLE_TABLE]: {
		resourceType: "AWS::Serverless::SimpleTable"
	}
}

export interface GlobalConfigItem {
	resourceType: string
	properties: Segment[]
}

export interface GlobalsConfig {
	[API]: GlobalConfigItem
	[FUNCTION]: GlobalConfigItem
	[SIMPLE_TABLE]: GlobalConfigItem
}

const KEYS = [API, FUNCTION, SIMPLE_TABLE]

export const isEmpty = (globalsConfig: GlobalsConfig): boolean => {
	return KEYS.every(key => globalsConfig[key].properties.length === 0)
}
