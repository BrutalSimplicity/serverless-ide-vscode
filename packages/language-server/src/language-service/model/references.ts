import { ASTNode } from "./../parser/json"

export enum ReferenceType {
	// getters
	FIND_IN_MAP = "FIND_IN_MAP",
	GET_ATT = "GET_ATT",
	// transformers
	SUB = "SUB",
	// ref/sub
	REF = "REF",
	// depends on
	DEPENDS_ON = "DEPENDS_ON"
}

export interface Reference {
	type: ReferenceType
	key: string
	node: ASTNode
	offset: number
}
