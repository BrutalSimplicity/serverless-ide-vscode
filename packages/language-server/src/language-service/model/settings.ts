export enum ValidationProvider {
	default = "default",
	"cfn-lint" = "cfn-lint"
}

export interface CFNLintExtensionSettings {
	path?: string
	appendRules?: string[]
	ignoreRules?: string[]
	overrideSpecPath?: string
}

export interface ExtensionSettings {
	serverlessIDE: {
		validationProvider?: ValidationProvider
		cfnLint?: CFNLintExtensionSettings
		validate?: boolean
		hover?: boolean
		completion?: boolean
	}
	http: {
		proxy: string
		proxyStrictSSL: boolean
	}
}

export interface CFNLintSettings {
	path: string
	appendRules: string[]
	ignoreRules: string[]
	overrideSpecPath?: string
}

export interface LanguageSettings {
	validate: boolean
	hover: boolean
	completion: boolean
	customTags: string[]
	validationProvider: ValidationProvider
	cfnLint: CFNLintSettings
	workspaceRoot: string
}

export const getDefaultLanguageSettings = (): LanguageSettings => ({
	validate: true,
	hover: true,
	completion: true,
	customTags: [],
	validationProvider: ValidationProvider["cfn-lint"],
	cfnLint: {
		path: "cfn-lint",
		appendRules: [],
		ignoreRules: []
	},
	workspaceRoot: ""
})
