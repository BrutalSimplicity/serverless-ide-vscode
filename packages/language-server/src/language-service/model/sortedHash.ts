export interface SerializedSortedHash<TItem> {
	sequence: string[]
	hash: {
		[key: string]: TItem
	}
}

export class SortedHash<TItem> {
	private sequence: string[]
	private hash: { [key: string]: TItem }

	constructor(
		defaultValue: SerializedSortedHash<TItem> = { sequence: [], hash: {} }
	) {
		this.sequence = defaultValue.sequence
		this.hash = defaultValue.hash
	}

	add(key: string, item: TItem) {
		if (this.hash[key]) {
			throw new Error(`Object with \`${key}\` already exists.`)
		}

		this.sequence.push(key)
		this.hash[key] = item
	}

	get(key: string): TItem | void {
		return this.hash[key]
	}

	contains(key: string): boolean {
		return key in this.hash
	}

	getList(): TItem[] {
		return this.map(item => item)
	}

	map<TValue>(callback: (item: TItem, index: number) => TValue): TValue[] {
		return this.sequence.map((key, index) => {
			return callback(this.hash[key], index)
		})
	}

	forEach(callback: (item: TItem, index: number) => void) {
		return this.sequence.forEach((key, index) => {
			return callback(this.hash[key], index)
		})
	}

	insertAtIndex(key: string, item: TItem, index: number) {
		this.sequence = [
			...this.sequence.slice(0, index),
			key,
			...this.sequence.slice(index)
		]
		this.hash[key] = item
	}

	remove(key: string) {
		const index = this.sequence.indexOf(key)

		if (index !== -1) {
			this.sequence.splice(index, 1)
			delete this.hash[key]
		}
	}

	serialize() {
		return {
			sequence: [...this.sequence],
			hash: {
				...this.hash
			}
		}
	}
}
