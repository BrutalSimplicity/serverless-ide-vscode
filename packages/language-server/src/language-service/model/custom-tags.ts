import { ReferenceType } from "./references"

export type TagKind = "sequence" | "scalar" | "mapping"

export interface CustomTag {
	type?: ReferenceType
	tag: string
	kind: TagKind
	propertyName: string
}

export const CUSTOM_TAGS: CustomTag[] = [
	{
		tag: "!Base64",
		kind: "sequence",
		propertyName: "Fn::Base64"
	},
	{
		tag: "!Cidr",
		kind: "sequence",
		propertyName: "Fn::Cidr"
	},
	// logical
	{
		tag: "!And",
		kind: "sequence",
		propertyName: "Fn::And"
	},
	{
		tag: "!If",
		kind: "sequence",
		propertyName: "Fn::If"
	},
	{
		tag: "!Not",
		kind: "sequence",
		propertyName: "Fn::Not"
	},
	{
		tag: "!Or",
		kind: "sequence",
		propertyName: "Fn::Or"
	},
	{
		tag: "!Equals",
		kind: "sequence",
		propertyName: "Fn::Equals"
	},
	// getters
	{
		type: ReferenceType.FIND_IN_MAP,
		tag: "!FindInMap",
		kind: "sequence",
		propertyName: "Fn::FindInMap"
	},
	{
		type: ReferenceType.GET_ATT,
		tag: "!GetAtt",
		kind: "scalar",
		propertyName: "Fn::GetAtt"
	},
	{
		tag: "!GetAZs",
		kind: "scalar",
		propertyName: "Fn::GetAZs"
	},
	{
		tag: "!ImportValue",
		kind: "mapping",
		propertyName: "Fn::ImportValue"
	},
	{
		tag: "!Join",
		kind: "sequence",
		propertyName: "Fn::Join"
	},
	{
		tag: "!Select",
		kind: "sequence",
		propertyName: "Fn::Select"
	},
	{
		tag: "!Split",
		kind: "sequence",
		propertyName: "Fn::Split"
	},
	{
		type: ReferenceType.SUB,
		tag: "!Sub",
		kind: "sequence",
		propertyName: "Fn::Sub"
	},
	{
		tag: "!Transform",
		kind: "mapping",
		propertyName: "Fn::Transform"
	},
	// ref
	{
		type: ReferenceType.REF,
		tag: "!Ref",
		kind: "scalar",
		propertyName: "Ref"
	},
	{
		type: ReferenceType.DEPENDS_ON,
		tag: "",
		kind: "sequence",
		propertyName: "DependsOn"
	}
]
