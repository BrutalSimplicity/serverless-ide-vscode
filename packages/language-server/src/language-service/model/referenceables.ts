import { PropertyASTNode } from "../parser/json"
import { ReferenceEntityType } from "./references"

export interface Referenceable {
	id: string
	node: PropertyASTNode
	entityType: ReferenceEntityType
}

export interface ReferenceablesHash {
	[key: string]: Referenceable
}

export type ReferenceableLookup = WeakMap<PropertyASTNode, Referenceable>

export interface Referenceables {
	hash: {
		[key in ReferenceEntityType]: ReferenceablesHash
	}
	lookup: ReferenceableLookup
}
