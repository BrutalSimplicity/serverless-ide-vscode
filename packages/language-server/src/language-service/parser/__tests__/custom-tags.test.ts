import { CUSTOM_TAGS, TagKind } from "./../../model/custom-tags"
import { parse as parseYaml } from ".."

const optionsMapping: { [key in TagKind]: string } = {
	scalar: "logicVariable",
	sequence: ["", "   - logicalVariable1", "   - logicalVariable2"].join("\n"),
	mapping: "{ val1: logicalVariable1, val2: logicalVariable2 }"
}

describe("custom tags parse", () => {
	describe("smoke tests", () => {
		const generateNode = (text: string) => {
			return parseYaml(text).root
		}

		CUSTOM_TAGS.forEach(tag => {
			describe(tag.type, () => {
				if (tag.type) {
					test(`${tag.type}`, () => {
						const text = [
							`property: ${tag.tag} `,
							optionsMapping[tag.kind]
						].join("")
						const node = generateNode(text)

						expect(node).toMatchSnapshot()
					})
				}
			})
		})
	})
})
