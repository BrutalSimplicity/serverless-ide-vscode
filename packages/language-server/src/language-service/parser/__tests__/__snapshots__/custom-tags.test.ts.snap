// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom tags parse smoke tests  undefined 1`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 42,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 42,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Base64",
          "tag": "!Base64",
        },
        "end": 42,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 39,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 23,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 21,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 2`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 40,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 40,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Cidr",
          "tag": "!Cidr",
        },
        "end": 40,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 37,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 21,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 19,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 3`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 39,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 39,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::And",
          "tag": "!And",
        },
        "end": 39,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 36,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 20,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 18,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 4`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 38,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 38,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::If",
          "tag": "!If",
        },
        "end": 38,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 35,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 19,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 17,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 5`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 39,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 39,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Not",
          "tag": "!Not",
        },
        "end": 39,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 36,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 20,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 18,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 6`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 38,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 38,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Or",
          "tag": "!Or",
        },
        "end": 38,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 35,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 19,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 17,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 7`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 42,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 42,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Equals",
          "tag": "!Equals",
        },
        "end": 42,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 39,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 23,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 21,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 8`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 31,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 31,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": StringASTNode {
        "customTag": Object {
          "kind": "scalar",
          "propertyName": "Fn::GetAZs",
          "tag": "!GetAZs",
        },
        "end": 31,
        "isKey": false,
        "location": "property",
        "parent": [Circular],
        "start": 18,
        "type": "string",
        "value": "logicVariable",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 9`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 73,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 73,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ObjectASTNode {
        "customTag": undefined,
        "end": 73,
        "location": "property",
        "parent": [Circular],
        "properties": Array [
          PropertyASTNode {
            "colonOffset": -1,
            "customTag": undefined,
            "end": 47,
            "key": StringASTNode {
              "customTag": undefined,
              "end": 29,
              "isKey": true,
              "location": "val1",
              "parent": [Circular],
              "start": 25,
              "type": "string",
              "value": "val1",
            },
            "location": null,
            "parent": [Circular],
            "start": 25,
            "type": "property",
            "value": StringASTNode {
              "customTag": undefined,
              "end": 47,
              "isKey": false,
              "location": "val1",
              "parent": [Circular],
              "start": 31,
              "type": "string",
              "value": "logicalVariable1",
            },
          },
          PropertyASTNode {
            "colonOffset": -1,
            "customTag": undefined,
            "end": 71,
            "key": StringASTNode {
              "customTag": undefined,
              "end": 53,
              "isKey": true,
              "location": "val2",
              "parent": [Circular],
              "start": 49,
              "type": "string",
              "value": "val2",
            },
            "location": null,
            "parent": [Circular],
            "start": 49,
            "type": "property",
            "value": StringASTNode {
              "customTag": undefined,
              "end": 71,
              "isKey": false,
              "location": "val2",
              "parent": [Circular],
              "start": 55,
              "type": "string",
              "value": "logicalVariable2",
            },
          },
        ],
        "start": 23,
        "type": "object",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 10`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 40,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 40,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Join",
          "tag": "!Join",
        },
        "end": 40,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 37,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 21,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 19,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 11`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 42,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 42,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Select",
          "tag": "!Select",
        },
        "end": 42,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 39,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 23,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 21,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 12`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 41,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 41,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Split",
          "tag": "!Split",
        },
        "end": 41,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 38,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 22,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 20,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests  undefined 13`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 71,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 71,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ObjectASTNode {
        "customTag": undefined,
        "end": 71,
        "location": "property",
        "parent": [Circular],
        "properties": Array [
          PropertyASTNode {
            "colonOffset": -1,
            "customTag": undefined,
            "end": 45,
            "key": StringASTNode {
              "customTag": undefined,
              "end": 27,
              "isKey": true,
              "location": "val1",
              "parent": [Circular],
              "start": 23,
              "type": "string",
              "value": "val1",
            },
            "location": null,
            "parent": [Circular],
            "start": 23,
            "type": "property",
            "value": StringASTNode {
              "customTag": undefined,
              "end": 45,
              "isKey": false,
              "location": "val1",
              "parent": [Circular],
              "start": 29,
              "type": "string",
              "value": "logicalVariable1",
            },
          },
          PropertyASTNode {
            "colonOffset": -1,
            "customTag": undefined,
            "end": 69,
            "key": StringASTNode {
              "customTag": undefined,
              "end": 51,
              "isKey": true,
              "location": "val2",
              "parent": [Circular],
              "start": 47,
              "type": "string",
              "value": "val2",
            },
            "location": null,
            "parent": [Circular],
            "start": 47,
            "type": "property",
            "value": StringASTNode {
              "customTag": undefined,
              "end": 69,
              "isKey": false,
              "location": "val2",
              "parent": [Circular],
              "start": 53,
              "type": "string",
              "value": "logicalVariable2",
            },
          },
        ],
        "start": 21,
        "type": "object",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests FIND_IN_MAP FIND_IN_MAP 1`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 45,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 45,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::FindInMap",
          "tag": "!FindInMap",
          "type": "FIND_IN_MAP",
        },
        "end": 45,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 42,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 26,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 24,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests GET_ATT GET_ATT 1`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 31,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 31,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": StringASTNode {
        "customTag": Object {
          "kind": "scalar",
          "propertyName": "Fn::GetAtt",
          "tag": "!GetAtt",
          "type": "GET_ATT",
        },
        "end": 31,
        "isKey": false,
        "location": "property",
        "parent": [Circular],
        "start": 18,
        "type": "string",
        "value": "logicVariable",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests REF REF 1`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 28,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 28,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": StringASTNode {
        "customTag": Object {
          "kind": "scalar",
          "propertyName": "Ref",
          "tag": "!Ref",
          "type": "REF",
        },
        "end": 28,
        "isKey": false,
        "location": "property",
        "parent": [Circular],
        "start": 15,
        "type": "string",
        "value": "logicVariable",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;

exports[`custom tags parse smoke tests SUB SUB 1`] = `
ObjectASTNode {
  "customTag": undefined,
  "end": 39,
  "location": null,
  "parent": null,
  "properties": Array [
    PropertyASTNode {
      "colonOffset": -1,
      "customTag": undefined,
      "end": 39,
      "key": StringASTNode {
        "customTag": undefined,
        "end": 8,
        "isKey": true,
        "location": "property",
        "parent": [Circular],
        "start": 0,
        "type": "string",
        "value": "property",
      },
      "location": null,
      "parent": [Circular],
      "start": 0,
      "type": "property",
      "value": ArrayASTNode {
        "customTag": Object {
          "kind": "sequence",
          "propertyName": "Fn::Sub",
          "tag": "!Sub",
          "type": "SUB",
        },
        "end": 39,
        "items": Array [
          StringASTNode {
            "customTag": undefined,
            "end": 36,
            "isKey": false,
            "location": 0,
            "parent": [Circular],
            "start": 20,
            "type": "string",
            "value": "logicalVariable1",
          },
        ],
        "location": "property",
        "parent": [Circular],
        "start": 18,
        "type": "array",
      },
    },
  ],
  "start": 0,
  "type": "object",
}
`;
